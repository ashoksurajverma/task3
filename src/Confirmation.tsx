import React, { useState, useEffect } from "react";
import "./confirmation.css";

const alertClasses = {
  primary: "alert alert-primary",
  secondary: "alert alert-secondary",
  success: "alert alert-success",
  error: "alert alert-danger",
  caution: "alert alert-warning",
  message: "alert alert-info",
  light: "alert alert-light",
  dark: "alert alert-dark",
};
interface IProps {
  message: string;
  type:
    | "primary"
    | "secondary"
    | "success"
    | "error"
    | "caution"
    | "message"
    | "light"
    | "dark";
  acceptLabel: string;
  declineLabel: string;
  accept: () => void;
  decline: () => void;
}

function Confirmation({
  message,
  type,
  accept,
  decline,
  acceptLabel,
  declineLabel,
}: IProps) {
  const [shouldRender, setShouldRender] = useState(true);

  const onClickHandle = (isCofirmation: boolean) => {
    if (isCofirmation) {
      accept();
    } else {
      decline();
    }
    setShouldRender(false);
  };

  const getClassName = () => {
    if (!type) return alertClasses.message;
    return alertClasses[type];
  };

  if (!shouldRender) return null;

  return (
    <div className={getClassName()}>
      <p>{message}</p>
      <div className="btn-container">
        <div onClick={() => onClickHandle(true)} className="btn btn-primary">
          {acceptLabel}
        </div>
        <div onClick={() => onClickHandle(false)} className="btn btn-danger">
          {declineLabel}
        </div>
      </div>
    </div>
  );
}

export default Confirmation;
