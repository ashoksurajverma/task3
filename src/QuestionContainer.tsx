import React, { useState } from "react";
import Confirmation from "./Confirmation";

interface IQuestion {
  question: string;
  answer: string;
}
function QuestionContainer({ question, answer }: IQuestion) {
  const [showAnswer, setShowAnswer] = useState(false);
  const [showConfirmation, setShowConfirmation] = useState(false);

  const accept = () => {
    console.log("User accept anotification");
    setShowAnswer(true);
  };

  const decline = () => {
    console.log("User decline anotification");
    setShowConfirmation(false);
  };

  const onHandleShowAns = () => {
    console.log(showAnswer);
    if (showAnswer && showConfirmation) return;
    if (showAnswer) {
      setShowAnswer(false);
    }
    setShowConfirmation(!showConfirmation);
  };

  return (
    <div className="container">
      {showConfirmation && (
        <Confirmation
          accept={accept}
          decline={decline}
          message="Reveal the answer?"
          type="message"
          acceptLabel="Yes Please"
          declineLabel="Not Yet"
        />
      )}
      <div className="question">{question}</div>
      <div
        className={`btn btn-primary show-answer ${
          showAnswer && showConfirmation ? "disabled" : ""
        }`}
        onClick={onHandleShowAns}
      >
        Show Answer
      </div>
      {showAnswer && (
        <div className="answer">
          <span className="answer-label">Ans: </span>
          {answer}
        </div>
      )}
    </div>
  );
}

export default QuestionContainer;
