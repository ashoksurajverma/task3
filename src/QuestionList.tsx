import React from "react";
import QuestionContainer from "./QuestionContainer";

interface IQuestion {
  question: string;
  answer: string;
}

interface IProps {
  questions: Array<IQuestion>;
}
function QuestionList({ questions }: IProps) {
  return (
    <div className="question-wrapper">
      {questions.map((question: IQuestion, index: number) => {
        return <QuestionContainer {...question} key={index} />;
      })}
    </div>
  );
}

export default QuestionList;
