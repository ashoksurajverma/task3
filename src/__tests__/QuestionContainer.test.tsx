import React from "react";
import { shallow, ShallowWrapper, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import QuestionList from "../QuestionList";
import QuestionContainer from "../QuestionContainer";
import Confirmation from "../Confirmation";

configure({ adapter: new Adapter() });

describe("Render QuestionContainer component", () => {
  let wrapper: ShallowWrapper<undefined, undefined>;
  const question =
    "What is the average the airspeed velocity of a (European) unladen swallow?";
  const answer = "11 meters per second";

  beforeEach(() => {
    wrapper = shallow(
      <QuestionContainer question={question} answer={answer} />
    );
  });

  it("should render without error", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should render container", () => {
    expect(wrapper.find(".container").length).toBe(1);
  });

  it("Should not render confirmation component initially", () => {
    expect(wrapper.children().find(Confirmation).length).toBe(0);
  });

  it("Should render confirmation component", () => {
    const showAnsBtn = wrapper.find(".show-answer");
    showAnsBtn.simulate("click");
    expect(wrapper.children().find(Confirmation).length).toBe(1);
  });

  it("get btns of confirmation component", () => {
    const showAnsBtn = wrapper.find(".show-answer");
    showAnsBtn.simulate("click");
    const confirmBtn = wrapper.find(".btn-primary");
    expect(confirmBtn.length).toBe(1);
  });
});
