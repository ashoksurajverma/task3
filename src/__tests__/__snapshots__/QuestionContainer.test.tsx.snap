// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render QuestionContainer component should render without error 1`] = `ShallowWrapper {}`;
